### PR TITLE
Add RPC endpoints for config entry operations

### PR DESCRIPTION
**Pull request by kyhavlov**
This PR adds RPC endpoints for config entries to be used in upcoming Connect work around centralized service config.

### CODE REVIEW

File: agent/structs/structs.go [hunk 1]
```diff
@@ -1140,6 +1140,82 @@ type IndexedNodeDump struct {
 	QueryMeta
 }
 
+// IndexedConfigEntries has its own encoding logic which differs from
+// ConfigEntryRequest as it has to send a slice of ConfigEntry.
+type IndexedConfigEntries struct {
+	Entries []ConfigEntry
+	QueryMeta
```
Comment (banks) [inline agent/structs/structs.go]: Let's just add `Kind` here and make it only support single type in one response.

File: agent/structs/config_entry.go [hunk 2]
```diff
@@ -25,6 +26,11 @@ type ConfigEntry interface {
 	Normalize() error
 	Validate() error
 
+	// VerifyReadACL and VerifyWriteACL return whether or not the given Authorizer
+	// has permission to read or write to the config entry, respectively.
+	VerifyReadACL(acl.Authorizer) bool
+	VerifyWriteACL(acl.Authorizer) bool
```
Comment (banks) [inline agent/structs/config_entry.go]: How do you feel about names like `CanRead(), CanWrite()`. I think this looks great but just the conditionals ended up not reading super cleanly to me:

```go
if rule != nil && !args.Entry.VerifyWriteACL(rule) {
```

Is kinda hard to parse and figure out the negations and then assume that "Verify" returns true on failure etc. Whereas this seems slightly clearer as it's an assertion:

```go
if rule != nil && !args.Entry.CanWrite(rule) {
```

It's minor since we shouldn't need to call these in too many other places now but just a thought.

File: agent/structs/structs.go [hunk 3]
```diff
@@ -1153,16 +1154,16 @@ func (c *IndexedConfigEntries) MarshalBinary() (data []byte, err error) {
 	bs := make([]byte, 128)
 	enc := codec.NewEncoderBytes(&bs, msgpackHandle)
 
-	// Encode kinds of entries first
+	// Encode length.
 	err = enc.Encode(len(c.Entries))
 	if err != nil {
 		return nil, err
 	}
```
Comment (banks) [inline agent/structs/structs.go]: We shouldn't need to encode length at all any more right - msgpack will do that as part of the slice serialisation and we only have a single kind value before the main struct now.

We can just decode into a nil `IndexedConfigEntries` and msgpack will figure out length and allocations etc.

Comment (kyhavlov) [reply agent/structs/structs.go]: I think we still need to fill in a dummy entry for each spot in the list with `MakeConfigEntry` - I couldn't find a way for it to work without populating the slice beforehand.
